torch/csrc/jit/passes/to_batch.cpp

```diff
@@ -1,63 +1,290 @@
 #include "torch/csrc/jit/passes/to_batch.h"
 #include "torch/csrc/jit/script/compiler.h"
+#include "torch/csrc/jit/tensor_conversions.h"
 
 namespace torch { namespace jit {
 
 std::unordered_map<std::string, std::shared_ptr<Graph>> ToBatch::batch_operator_table;
 
-void ToBatch::toBatch(Block* block, Block* res_block) {
-  // change inputs of a graph - expand tensor to {data, mask, dims}
+// eg: "a.1" -> {"a", "1"}; "a" -> {"a"}
+std::vector<std::string> ToBatch::get_name(std::string name) {
+  auto last_dot_pos = name.find_last_of('.');
+  if (last_dot_pos != std::string::npos && last_dot_pos + 1 != name.size()) {
+    if (name.find_first_not_of("0123456789", last_dot_pos + 1) == std::string::npos) {
+      auto suffix = name.substr(last_dot_pos + 1);
+      std::string name_base = name.substr(0, last_dot_pos);
+      return {name_base, suffix};
+    }
+  }
+  return {name};
+}
+
+// replace aten operator node with BatchTensor operator graph
+void ToBatch::visitAten(Node* n, Block* block, Block* res_block, std::unordered_map<std::string, Value*>& var_map){
+  if(n->outputs().size() > 1){
+    throw std::runtime_error("Cannot process multiple assignment");
+  }
+  auto batch_graph = batch_operator_table.at(n->kind().toUnqualString());
+  std::vector<Value*> new_inputs;
+  for(Value *input : n->inputs()){
+    if(batch_map.find(input) != batch_map.end()){
+      auto new_input = batch_map.at(input);
+      new_inputs.insert(new_inputs.end(), new_input.begin(), new_input.end());
+    }
+    else{
+      throw std::runtime_error("NYI: non-tensor input for aten operator is not supported yet");
+    }
+  }
+  auto outputs = script::inlineCallTo(*res_block->owningGraph(), *batch_graph, new_inputs);
+
+  // do update on assignment
+  auto name_base = get_name(n->output()->uniqueName())[0];
+  if(var_map.find(name_base) != var_map.end()){
+    std::vector<Value*> inputs(batch_map.at(var_map.at(name_base)));
+    inputs.insert(inputs.end(), outputs.begin(), outputs.end());
+    outputs = script::inlineCallTo(*res_block->owningGraph(), *batch_operator_table.at("update"), inputs);
+  }
+  // Assume all outputs from inlined operator implementation are in the triple form.
+  for(size_t i = 0; i < n->outputs().size(); i++){
+    auto output = n->outputs()[i];
+    batch_map[output] = std::vector<Value*>(outputs.begin() + i * 3, outputs.begin() + i * 3 + 3);
+    if(output->hasUniqueName()){
+      var_map[get_name(output->uniqueName())[0]] = output;
+    }
+  }
+}
+
+// clone prim::Constant to new graph
```

Comment:
I apply a batching transformation to the output of `prim::NumToTensor`. So if there is a `prim::NumToTensor` following `prim::Constant`, it will be transformed.

Reply:
Please copy that into a comment here :D